### PR TITLE
fix: cookie value comparison

### PR DESCRIPTION
**Pull request by MoumitaM**
## PR Description

I have fixed the current behavior of cookie value comparison between the actual value and the value set from the server side.

## Linear task (optional)

[Linear task link](https://linear.app/rudderstack/issue/SDK-1930/fix-cookie-value-comparison)

## Cross Browser Tests

Please confirm you have tested for the following browsers:

- [x] Chrome
- [ ] Firefox
- [ ] IE11

## Sanity Suite

- [ ] All sanity suite test cases pass locally

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved cookie value comparison in user session management to handle complex objects properly, reducing potential errors.

- **Tests**
  - Added and mocked a new utility function in user session manager tests to ensure accurate validation of complex objects.

### CODE REVIEW

File: packages/analytics-js/src/components/userSessionManager/UserSessionManager.ts [hunk 1]
```diff
@@ -361,7 +361,9 @@ class UserSessionManager implements IUserSessionManager {
             cookieData.forEach(each => {
               const cookieValue = store?.get(each.name);
               if (each.value) {
-                if (cookieValue !== each.value) {
+                const before = stringifyWithoutCircular(each.value, false, []);
+                const after = stringifyWithoutCircular(cookieValue, false, []);
+                if (after !== before) {
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply packages/analytics-js/src/components/userSessionManager/UserSessionManager.ts:366]: Not necessary

**Review by saikumarrs — COMMENTED** [review]
You are merely checking if the stringify utility is invoked. However, that doesn't guarantee anything.
Instead, test the scenario where the comparison can pass and fail.

**Comment by MoumitaM** [comment]
> You are merely checking if the stringify utility is invoked. However, that doesn't guarantee anything. Instead, test the scenario where the comparison can pass and fail.

Test cases for that is already present

**Comment by saikumarrs** [comment]
> > You are merely checking if the stringify utility is invoked. However, that doesn't guarantee anything. Instead, test the scenario where the comparison can pass and fail.
> 
> Test cases for that is already present

I'm good as long as those test cases also test cookies like user traits and session information (whose values are objects) as that's the motivation for this change.

**Review by saikumarrs — APPROVED** [review]

**Pull request merged**